File: python/src/wh40kdc/data/affordability.py
"""candidate_affordability — given the units already in a list and a points
budget, price the cheapest next copy of each candidate unit and flag whether it
still fits. Powers the list-builder's "sort cheapest-first / grey out the
unaffordable" catalog affordance, and is exposed as a cross-impl primitive so
the maths is pinned by ``conformance/affordability/``.

The cost of "one more copy" is ordinal-aware: 11e prices some datasheets by army
ordinal (see :func:`base_unit_points`), so the next copy of a datasheet you
already field twice may cost more than the first. ``nextCopyCost`` is the
cheapest *entry point* — the minimum over the unit's points tiers of
``base_unit_points(unit, tier.models, next_ordinal)`` — i.e. taking it at its
smallest legal size, at the ordinal it would enter the army.

Python mirror of ``tools/src/data/affordability.ts`` /
``crates/wh40kdc/src/data/affordability.rs`` / ``go/affordability.go``.
"""

from __future__ import annotations

import math
from typing import Any

from wh40kdc.data.battle_sizes import points_limit_for_battle_size
from wh40kdc.data.dataset import Dataset
from wh40kdc.data.pricing import host_points_tiers, host_unit_points


def _cheapest_next_copy(
    unit: dict[str, Any],
    next_ordinal: int,
    host_faction: dict[str, Any] | None,
) -> int:
    """The cheapest cost to field one more copy of ``unit`` at ``next_ordinal``,
    over the tiers in effect for ``host_faction``'s army."""
    tiers = host_points_tiers(unit, host_faction)
    if not tiers:
        return 0
    best = math.inf
    for t in tiers:
        cost = host_unit_points(unit, t["models"], next_ordinal, host_faction)
        if cost < best:
            best = cost
    return 0 if best == math.inf else int(best)


def candidate_affordability(spec: dict[str, Any], dataset: Dataset) -> list[dict[str, Any]]:
    """Price the cheapest next copy of each candidate and flag affordability.

    ``spec`` carries ``faction_id`` (str | None), ``battle_size`` (str | None),
    ``points_limit_override`` (int | None), ``units`` (list of dicts with
    ``unit_id``/``model_count``/optional ``enhancement_id``), and optional
    ``candidate_unit_ids`` (list of ids, else every unit in ``faction_id``).

    Returns one dict ``{"unitId", "nextCopyCost", "affordable"}`` per candidate
    that resolves in the dataset, sorted ascending by ``(nextCopyCost, unitId)``
    — deterministic for conformance.
    """
    faction_id = spec.get("faction_id")

    def resolve(unit_id: str) -> Any:
        if not unit_id:
            return None
        if faction_id:
            scoped = dataset.units.get_in_faction(unit_id, faction_id)
            if scoped is not None:
                return scoped
        return dataset.units.get_any(unit_id)

    # Running total of the current list (ordinal-aware) + enhancement costs.
    # Host-aware: foreign units with an ``allied_points`` entry for this army
    # price from that entry (see ``host_points_tiers``).
    host_faction_view = dataset.factions.get(faction_id) if faction_id else None
    host_faction = host_faction_view.raw if host_faction_view else None
    ordinals: dict[str, int] = {}
    spent = 0
    for u in spec.get("units") or []:
        unit_id = u.get("unit_id") or ""
        view = resolve(unit_id)
        if view is None:
            continue
        ordinal = ordinals.get(unit_id, 0) + 1
        ordinals[unit_id] = ordinal
        spent += host_unit_points(view.raw, u.get("model_count") or 0, ordinal, host_faction)
        enhancement_id = u.get("enhancement_id")
        if enhancement_id:
            enh = dataset.enhancements.get(enhancement_id)
            spent += (enh or {}).get("cost") or 0

    override = spec.get("points_limit_override")
    limit = (
        override if override is not None else points_limit_for_battle_size(spec.get("battle_size"))
    )
    remaining = math.inf if limit is None else limit - spent

    candidate_ids = spec.get("candidate_unit_ids")
    if candidate_ids is None:
        candidate_ids = [v.id for v in dataset.units.by_faction(faction_id)] if faction_id else []

    out: list[dict[str, Any]] = []
    for unit_id in candidate_ids:
        view = resolve(unit_id)
        if view is None:
            continue
        next_ordinal = ordinals.get(unit_id, 0) + 1
        next_copy_cost = _cheapest_next_copy(view.raw, next_ordinal, host_faction)
        out.append(
            {
                "unitId": view.id,
                "nextCopyCost": next_copy_cost,
                "affordable": next_copy_cost <= remaining,
            }
        )
    out.sort(key=lambda c: (c["nextCopyCost"], c["unitId"]))
    return out

File: go/affordability.go
package wh40kdc

import (
	"math"
	"sort"
)

// candidate_affordability — given the units already in a list and a points
// budget, price the cheapest next copy of each candidate unit and flag whether
// it still fits. Go mirror of tools/src/data/affordability.ts.

// affordabilityUnit is one unit already in the list (fixes the running total +
// per-datasheet ordinals). enhancementID is "" for none.
type affordabilityUnit struct {
	unitID        string
	modelCount    int
	enhancementID string
}

// affordabilitySpec is the compact input shared by the runner op.
type affordabilitySpec struct {
	factionID           string
	battleSize          string
	pointsLimitOverride *int
	units               []affordabilityUnit
	candidateUnitIDs    []string
	candidateProvided   bool
}

// cheapestNextCopy is the cheapest cost to field one more copy of view at army
// ordinal nextOrdinal (minimum over the tiers in effect for hostFaction's army).
func cheapestNextCopy(view *UnitView, nextOrdinal int, hostFaction map[string]any) int {
	tiers := hostPointsTiers(view.Raw, hostFaction)
	if len(tiers) == 0 {
		return 0
	}
	min := math.MaxInt
	for _, tAny := range tiers {
		t, _ := asMap(tAny)
		cost := hostUnitPoints(view.Raw, asInt(t["models"]), nextOrdinal, hostFaction)
		if cost < min {
			min = cost
		}
	}
	if min == math.MaxInt {
		return 0
	}
	return min
}

// candidateAffordability prices the cheapest next copy of each candidate and
// flags affordability against the remaining budget. Returns one object per
// candidate that resolves in the dataset, sorted ascending by
// (nextCopyCost, unitId). Mirror of the TS reference.
func candidateAffordability(spec affordabilitySpec, ds *Dataset) []map[string]any {
	resolve := func(unitID string) *UnitView {
		if unitID == "" {
			return nil
		}
		if spec.factionID != "" {
			if uv, ok := ds.Units.GetInFaction(unitID, spec.factionID); ok {
				return uv
			}
		}
		if uv, ok := ds.Units.GetAny(unitID); ok {
			return uv
		}
		return nil
	}

	// Running total of the current list (ordinal-aware) + enhancement costs.
	// Host-aware: foreign units with an allied_points entry for this army price
	// from that entry (see hostPointsTiers).
	var hostFaction map[string]any
	if spec.factionID != "" {
		if f, ok := ds.Factions.Get(spec.factionID); ok {
			hostFaction = f.Raw
		}
	}
	ordinals := map[string]int{}
	spent := 0
	for _, u := range spec.units {
		view := resolve(u.unitID)
		if view == nil {
			continue
		}
		ord := ordinals[u.unitID] + 1
		ordinals[u.unitID] = ord
		spent += hostUnitPoints(view.Raw, u.modelCount, ord, hostFaction)
		if u.enhancementID != "" {
			if eAny, ok := ds.Enhancements.Get(u.enhancementID); ok {
				spent += asInt(eAny.(map[string]any)["cost"])
			}
		}
	}

	hasLimit := false
	limit := 0
	if spec.pointsLimitOverride != nil {
		limit = *spec.pointsLimitOverride
		hasLimit = true
	} else if l, ok := pointsLimitForBattleSize(spec.battleSize); ok {
		limit = l
		hasLimit = true
	}
	remaining := 0
	if hasLimit {
		remaining = limit - spent
	}

	// Candidate set: explicit list, else every unit in the faction.
	candidateIDs := spec.candidateUnitIDs
	if !spec.candidateProvided {
		candidateIDs = nil
		if spec.factionID != "" {
			for _, v := range ds.Units.ByFaction(spec.factionID) {
				candidateIDs = append(candidateIDs, v.ID())
			}
		}
	}

	out := []map[string]any{}
	for _, unitID := range candidateIDs {
		view := resolve(unitID)
		if view == nil {
			continue
		}
		nextOrdinal := ordinals[unitID] + 1
		nextCopyCost := cheapestNextCopy(view, nextOrdinal, hostFaction)
		affordable := !hasLimit || nextCopyCost <= remaining
		out = append(out, map[string]any{
			"unitId":       view.ID(),
			"nextCopyCost": nextCopyCost,
			"affordable":   affordable,
		})
	}
	sort.SliceStable(out, func(i, j int) bool {
		ci, cj := out[i]["nextCopyCost"].(int), out[j]["nextCopyCost"].(int)
		if ci != cj {
			return ci < cj
		}
		return out[i]["unitId"].(string) < out[j]["unitId"].(string)
	})
	return out
}
